docs/csharp/language-reference/compiler-messages/cs8352.md

```diff
@@ -0,0 +1,120 @@
+---
+description: "Compiler Error CS8352"
+title: "Compiler Error CS8352"
+ms.date: 12/05/2023
+f1_keywords:
+  - "CS8352"
+helpviewer_keywords:
+  - "CS8352"
+---
+# Compiler Error CS8352
+
+Cannot use variable 'variable' in this context because it may expose referenced variables outside of their declaration scope.
+
+This error indicates potential undefined behavior when there's a risk of accessing a stack allocated variable after it has already been deleted from the stack.
+
+## To correct this error
+
+Ways to avoid `CS8352` are, for example:
+
+* Redesign the architecture of your solution regarding the way objects are allocated in the memory and the memory management in your implementation.
+* Use `in` keyword and create the object in wider scope
+* Use `new` keyword to create your object on a heap instead of a stack
+* Use `ref` and `unsafe` keyword to notify the compiler of intentional object passing
+
+## Example
+
+The following sample generates CS8352:
+
+```csharp
+class Program
+{
+    static public Span<int> CreateSpanWithValue(int size, int value)
+    {
+        Span<int> localSpan = stackalloc int[size];
+        localSpan[0] = value;
+        return localSpan;    // CS8352
+    }                        // localSpan object is deleted at this point
+
+    static public void Main()
+    {
+        Span<int> localSpan = CreateSpanWithValue(5, 10);
+        localSpan[2] = 14;
+        foreach(var item in localSpan)
+        {
+            Console.WriteLine(item.ToString());  // Reading from deleted object
+        }
+    }
+    // Example of output:
+    // 284945320
+    // 149
+    // 149
+    // 1369466325
+    // 284945320
+}
+```
+
+The following sample uses `in` keyword and creates the object in wider scope:
+
+```csharp
+class Program
+{
+    static public Span<int> FillSpanWithValue(in Span<int> span, int value)
+    {
+        span[0] = value;
+    }
+
+    static public void Main()
+    {
+        Span<int> localSpan = stackalloc int[5];
+        FillSpanWithValue(in localSpan, 10);
+        localSpan[2] = 14;
+        foreach(var item in localSpan)
+        {
+            Console.WriteLine(item.ToString());
+        }
+    }
+    // Output:
+    // 10
+    // 0
+    // 14
+    // 0
+    // 0
+}
+```
+
+The following sample uses the `new` keyword to create the object on heap, instead of stack:
+
+```csharp
+class Program
+{
+    static public Span<int> CreateSpanWithValue(int size, int value)
+    {
+        Span<int> localSpan = new int[size];
+        localSpan[1] = value;
+        return localSpan;
+    }
```

Comment:
I don't know if I follow - the two last examples are to present the bullet points of how to deal with the error.
So yes, they do highlight the concern but at the same time show a working solution.
Can you let me know what is your opinion of what should be done with those two examples?

Reply:
This is a complicated part of the language. (I had several meetings with the compiler devs before I felt I understood it well enough to write about it).

Effectively, you can think of `Span` as a wrapper around a pointer & a length. The static analysis ensures that the pointer refers to valid memory.

Let's start with the first sample:

```csharp
static public Span<int> CreateSpanWithValue(int size, int value)
{
    Span<int> localSpan = stackalloc int[size];
    localSpan[0] = value;
    return localSpan;    // CS8352
}
```

The reason for the error is that `localSpan` refers to memory that's on the stack, and only valid inside the method `CreateSpanWithValue`. That's because of the `stackalloc` expression to force the memory onto the local stack. The memory referred to by `localSpan` is reclaimed as part of unwinding the stack upon the `return` statement.

The second example moves the allocation up one level of method call:

```csharp
Span<int> localSpan = stackalloc int[5];
FillSpanWithValue(in localSpan, 10);
```

The reason this fixes the error is because `localSpan` is initialized to stack memory in the calling method, not the called method. That stack storage is still valid, because it was allocated in the `Main` method. The `in` modifier isn't necessary for the change to correct the error. The important difference is that the `stackalloc` expression is one level up the call stack.

The third example makes a different change:

```csharp
 Span<int> localSpan = new int[size];
```

That fixes the error because `localSpan` now refers to a `System.Array` object that was allocated on the heap. That array is still reachable after the method exists. The span still refers to valid memory.

This is a really complicated part of the language. The strict rules are discussed in these two sections of the standard:

- [Reference variables and returns](https://learn.microsoft.com/dotnet/csharp/language-reference/language-specification/variables#97-reference-variables-and-returns)
- [ref struct modifier](https://learn.microsoft.com/dotnet/csharp/language-reference/language-specification/structs#1623-ref-modifier)

Finally, [this feature spec](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-11.0/low-level-struct-improvements#examples-and-notes) has a number of interesting examples that demonstrate why the rules work the way they do. There are some serious mind-benders in there.
 

Reply:
> This is a complicated part of the language. (I had several meetings with the compiler devs before I felt I understood it well enough to write about it).
> 
> Effectively, you can think of `Span` as a wrapper around a pointer & a length. The static analysis ensures that the pointer refers to valid memory.

Yes, I agree and I'm aware of this.

> Let's start with the first sample:
> 
> ```cs
> static public Span<int> CreateSpanWithValue(int size, int value)
> {
>     Span<int> localSpan = stackalloc int[size];
>     localSpan[0] = value;
>     return localSpan;    // CS8352
> }
> ```
> 
> The reason for the error is that `localSpan` refers to memory that's on the stack, and only valid inside the method `CreateSpanWithValue`. That's because of the `stackalloc` expression to force the memory onto the local stack. The memory referred to by `localSpan` is reclaimed as part of unwinding the stack upon the `return` statement.

Exactly! This is my understanding of the reasons of the error as well.

> The second example moves the allocation up one level of method call:
> 
> ```cs
> Span<int> localSpan = stackalloc int[5];
> FillSpanWithValue(in localSpan, 10);
> ```
> 
> The reason this fixes the error is because `localSpan` is initialized to stack memory in the calling method, not the called method. That stack storage is still valid, because it was allocated in the `Main` method.

I agree. This was my thinking and intentions what I wanted to present when implementing the sample.

> The `in` modifier isn't necessary for the change to correct the error. The important difference is that the `stackalloc` expression is one level up the call stack.

This is something I wasn't aware of. Can you let me know if this is the mistake of mine you would like to emphasize?

> The third example makes a different change:
> 
> ```cs
>  Span<int> localSpan = new int[size];
> ```
> 
> That fixes the error because `localSpan` now refers to a `System.Array` object that was allocated on the heap. That array is still reachable after the method exists. The span still refers to valid memory.

Yes, this was my intention to show as another way of dealing with the error.

> This is a really complicated part of the language. The strict rules are discussed in these two sections of the standard:
> 
>     * [Reference variables and returns](https://learn.microsoft.com/dotnet/csharp/language-reference/language-specification/variables#97-reference-variables-and-returns)
> 
>     * [ref struct modifier](https://learn.microsoft.com/dotnet/csharp/language-reference/language-specification/structs#1623-ref-modifier)
> 
> 
> Finally, [this feature spec](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-11.0/low-level-struct-improvements#examples-and-notes) has a number of interesting examples that demonstrate why the rules work the way they do. There are some serious mind-benders in there.

Alright, thanks for all the sources 👍 

@BillWagner What I'm trying to understand is what exactly is missing, or is wrong with those two samples. Or maybe the whole approach I took is incorrect?
Thing is that I have the same understanding of root cause of CS8352 as you describe it is, and I thought I described/presented that root cause well. If not then can you point me some mistakes so I can correct those, or comment/decide regarding the whole approach?
Currently my guess, rather than understanding, of your comments to those two samples is that they do not show exactly what they should?

Reply:
@BartoszKlonowski 

> What I'm trying to understand is what exactly is missing, or is wrong with those two samples. Or maybe the whole approach I took is incorrect?

This is getting close. My main concern is that mentioning the `in` modifier will confuse readers. The `in` modifier doesn't fix the issue. The sharplab example in my earlier comment is your second sample, without the `in` modifier. The second sample works *not* because of the `in` modifier, but because the storage is allocated farther up the call stack. 

I'd do two things, and then this is ready:

1. Remove the `in` modifier from the second sample. It's not wrong, but it's not relevant.
1. In the text on the second and third samples, emphasize that the fix is for the Span to refer to memory that is still valid: either allocated on the heap, or on the stack in the calling context.

Again, I really appreciate all the work you've done here. This is an area that causes a lot of confusion for developers, and this will help once merged.